libbeat/esleg/eslegclient/config.go

```diff
@@ -0,0 +1,78 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package eslegclient
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/transport/tlscommon"
+)
+
+type config struct {
+	Hosts    []string          `config:"hosts" validate:"required"`
+	Protocol string            `config:"protocol"`
+	Path     string            `config:"path"`
+	Params   map[string]string `config:"parameters"`
+	Headers  map[string]string `config:"headers"`
+
+	TLS *tlscommon.Config `config:"ssl"`
+
+	ProxyURL     string `config:"proxy_url"`
+	ProxyDisable bool   `config:"proxy_disable"`
+
+	Username string `config:"username"`
+	Password string `config:"password"`
+	APIKey   string `config:"api_key"`
+
+	CompressionLevel int           `config:"compression_level" validate:"min=0, max=9"`
+	EscapeHTML       bool          `config:"escape_html"`
+	Timeout          time.Duration `config:"timeout"`
+}
+
+var (
+	defaultConfig = config{
+		Protocol:         "",
+		Path:             "",
+		ProxyURL:         "",
+		ProxyDisable:     false,
+		Params:           nil,
+		Username:         "",
+		Password:         "",
+		APIKey:           "",
+		Timeout:          90 * time.Second,
+		CompressionLevel: 0,
+		EscapeHTML:       false,
+		TLS:              nil,
+	}
+)
```

Comment:
config can store pointers and `map[string]interface{}`. I'd sleep a little better if we'd turn these template configurations into functions:

```
func defaultConfig() config {
  ...
}
```

This guarantees that we can never share pointers/references by accident.

Reply:
Fixed in 63664e5f4.